File: Lista_1/C/2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main()
{
    // função que verifica palíndromos (iterativa) em C (quase identica à versão em python)
    int verificaPalindromo(char palavra[]){
        int tamanho = strlen(palavra);
        for(int i = 0; i < tamanho / 2; i++){
            if(palavra[i] != palavra[tamanho - 1 - i]){
                return -1;
            }
        }
        return 1;
    }
    
    // falso teste
    char palavra[] = "arkara";
    printf("%d\n", verificaPalindromo(palavra));


    // versão recursiva que identifica se é palindromo
    int verificaPalindromoRec(char palavra[]){
        int tamanho = strlen(palavra);
        
        int verificaPRAux(int inicio, int fim){
            if(inicio >= fim){
                return 1;
            }
            
            if(palavra[inicio] != palavra[fim]){
                return -1;
            }
            
            verificaPRAux(inicio + 1, fim - 1);
        }
        
        verificaPRAux(0, tamanho - 1);
    }

    char palavra2[] = "elele";
    printf("%d", verificaPalindromoRec(palavra2));
    return 0;
}

File: Lista_1/Python/Python/2.py
#versão iterativa para verificar palindromos

palavra = "arara"
def verificaPalindromo(palavra):
    # considera que todas as letras são maiusculas ou minusculas
    # caso sejam diferentes, usar: palavra = palavra.lower();
    tamanho = len(palavra)
    i = 0
    for i in range(tamanho // 2): # percorre somente a metade, pois a outra metade deve ser igual
        if palavra[i] != palavra[tamanho - 1 - i]:
            return False
    return True

verificacao = verificaPalindromo(palavra)
print(verificacao)

#versao recursiva para identificar palindromos
def verificaPalindromoRec(palavra, indice = 0):
    if len(palavra) <= 1:
        return True
    elif palavra[0] != palavra[-1]:  # em python, uso de indices negativos se refere ao final da string para o começo
        return False
    else:
        return verificaPalindromoRec(palavra[1:-1])  #recursivamente passa a fatia entre o primeiro e o ultimo

verificacao2 = verificaPalindromoRec(palavra)
print(verificacao2)


def verificaPalindromoUL(palavra):
    return palavra == palavra[::-1] # percorre a string, de trás para frente, comparando os caracteres

verificacao3 = verificaPalindromoUL(palavra)
print(verificacao3)